Bug fix for hosts & vms not returned

iButcat opened the pull request:
Also added Tu's recommandation from his fork after reviewing the changes. 

Display of vms & hosts is now working as shown in the screenshot.

![Screenshot 2025-04-01 at 13 00 20](https://github.com/user-attachments/assets/c4c69a76-9954-4d1b-9b8e-60e6aef282b4)

iButcat committed:
fix bug where host and vm wasn't returned.

iButcat committed:
apply Tu's recommandation.

iButcat committed:
use throw instead of write error.

dinhngtu commented on formats/host.ps1xml:
```diff
@@ -1,11 +1,10 @@
     <ViewDefinitions>
         <View>
-            <Name>XoPowershell.Host</Name>
+            <Name>XoHost</Name>
```
Please make sure the type names are consistent across all PSObject types. Also, why the change?

dinhngtu commented on formats/vm.ps1xml:
```diff
@@ -3,17 +3,14 @@
     <ViewDefinitions>
         <View>
-            <Name>XoPowershell.Vm</Name>
+            <Name>XoVm</Name>
```
Same as above.

dinhngtu commented on src/host.ps1:
```diff
@@ -1,43 +1,73 @@
-$script:XO_HOST_FIELDS = "uuid,name_label,power_state,memory,address,hostname,version,productBrand"
+$script:XO_HOST_FIELDS = "id,uuid,name_label,name_description,power_state,memory,address,hostname,version,productBrand,build,cpus,startTime,tags,bios_strings,software_version,cpu_info,CPUs"
```
Why both `id` and `uuid`, aren't they the same? Let's choose one and stick with it.

dinhngtu commented on src/host.ps1:
```diff
@@ -1,43 +1,73 @@
     .DESCRIPTION
         Convert a host object from the API to a PowerShell object with proper properties and types.
-    .PARAMETER InputObject
+    .PARAMETER HostData
```
`InputObject` is idiomatic.

dinhngtu commented on src/host.ps1:
```diff
@@ -1,43 +1,73 @@
-            Build = $InputObject.software_version.build_number
-            Version = $InputObject.software_version.product_version
-            ProductBrand = $InputObject.software_version.product_brand
+    if ($HostData -is [string]) {
```
Is there any case that `Invoke-RestMethod` returns a JSON string?

dinhngtu commented on src/host.ps1:
```diff
@@ -1,43 +1,73 @@
+        }
+    }
+    
+    Write-Verbose "Processing host data: name_label=$($HostData.name_label), id=$($HostData.id), uuid=$($HostData.uuid), address=$($HostData.address)"
```
`$HostData | Out-String | Write-Verbose`

dinhngtu commented on src/host.ps1:
```diff
@@ -1,43 +1,73 @@
+    
+    Write-Verbose "Processing host data: name_label=$($HostData.name_label), id=$($HostData.id), uuid=$($HostData.uuid), address=$($HostData.address)"
+
+    $hostObj = [PSCustomObject]@{
```
Why build a new object instead of reusing `$HostData`? Also, this should be in `process{}`.

dinhngtu commented on src/host.ps1:
```diff
@@ -1,43 +1,73 @@
+    }
+
+    if ($HostData.ContainsKey('cpus') -and $HostData['cpus'] -and $HostData['cpus'].ContainsKey('cores')) {
+        $hostObj.Cpus = $HostData['cpus']['cores']
```
When does `CPUs.cores` exist? My JSON output only has `CPUs.max` and `CPUs.number`.

dinhngtu commented on src/host.ps1:
```diff
@@ -1,43 +1,73 @@
+
+    if ($HostData.ContainsKey('memory') -and $HostData['memory']) {
+        $memory = $HostData['memory']
+        if ($memory.ContainsKey('size')) {
```
You should test `$memory['size'] -eq $null` directly.

dinhngtu commented on src/host.ps1:
```diff
@@ -1,43 +1,73 @@
+        if ($memory.ContainsKey('size')) {
+            $hostObj.MemoryTotal = [math]::Round($memory['size'] / 1GB, 2)
+        }
+        if ($memory.ContainsKey('usage')) {
```
Does `$HostData['memory']['usage']` exist? Mine doesn't show any `usage` key in the JSON response even with Xen tools installed. Is it a new value from XO?

dinhngtu commented on src/host.ps1:
```diff
@@ -1,43 +1,73 @@
-        return $resultObject
     }
+
+    if ($hostObj.PowerState -and $hostObj.PowerState -ne "Unknown") {
```
There's no need to touch existing fields coming from XO. The PSObject view columns should only contain fields for things not immediately visible from the JSON object's top level.

dinhngtu commented on src/host.ps1:
```diff
@@ -75,88 +108,89 @@ function Get-XoHost {
         [Parameter(ParameterSetName = "Filter")]
         [Parameter(ParameterSetName = "All")]
-        [int]$Limit
+        [int]$Limit = 25
```
This wants to be presettable per-session somehow.

dinhngtu commented on src/host.ps1:
```diff
@@ -75,88 +108,89 @@ function Get-XoHost {
+                $uri = "$script:XoHost/rest/v0/hosts/$id"
+                Write-Verbose "Getting host with ID $id from $uri"
+                
+                $hostData = Invoke-RestMethod -Uri $uri @script:XoRestParameters
```
I just noticed, where did the request body go?

dinhngtu commented on src/host.ps1:
```diff
@@ -75,88 +108,89 @@ function Get-XoHost {
+        $allHostsResponse = Invoke-RestMethod -Uri $uri @script:XoRestParameters
+        Write-Verbose "API response received with $($allHostsResponse.Count) total hosts"
+
+        $hostsToProcess = $allHostsResponse
+        if ($Limit -gt 0 -and $allHostsResponse.Count -gt $Limit) {
+            $hostsToProcess = $allHostsResponse[0..($Limit-1)]
+            Write-Verbose "Processing first $Limit hosts out of $($allHostsResponse.Count) total"
+        } else {
+            Write-Verbose "Processing all $($allHostsResponse.Count) hosts"
+        }
```
Still need to pass limit parameter to API request.

dinhngtu commented on src/host.ps1:
```diff
@@ -75,88 +108,89 @@ function Get-XoHost {
+                $hostId = $matches[1]
+                try {
+                    Write-Verbose "Fetching host details for ID $hostId ($($processedCount+1) of $($hostsToProcess.Count))"
+                    $hostDetail = Invoke-RestMethod -Uri "$script:XoHost/rest/v0/hosts/$hostId" @script:XoRestParameters
```
Same; where did the request body go?

dinhngtu commented on src/host.ps1:
```diff
@@ -1,43 +1,73 @@
+    return $hostObj
 }
 
 function Get-XoHost {
```
Will doing everything in `end{}` break pipelines? Please stick with one cmdlet structure in all of your request (`Get-Xo...`) cmdlets.

dinhngtu commented on src/host.ps1:
```diff
@@ -75,88 +108,89 @@ function Get-XoHost {
-                                            $hostDetail = $hostDetail | ConvertFrom-Json -AsHashtable
-                                        }
-                                        
-                                        ConvertTo-XoHostObject $hostDetail
-                                    }
-                                }
-                                else {
-                                    Write-Warning "Failed to extract valid ID from URL: $hostUrl"
-                                }
-                            }
-                        } catch {
-                            Write-Warning "Failed to retrieve host details from URL: $hostUrl. Error: $_"
+            if ($cleanUrl -match "/hosts/([0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12})") {
```
Why the trim and match? Also, you can already request all host-relevant fields with one API call, why make one request per host?

dinhngtu commented on src/server.ps1:
```diff
@@ -111,137 +116,25 @@ function Get-XoServer {
-        }
-    }
-}
-
-function Disable-XoServer {
-    <#
-    .SYNOPSIS
-        Disable a server in Xen Orchestra.
-    .DESCRIPTION
-        Disables a server in Xen Orchestra. Disabled servers are not accessed by Xen Orchestra.
-    .PARAMETER ServerUuid
-        The UUID of the server to disable.
-    .EXAMPLE
-        Disable-XoServer -ServerUuid "12345678-abcd-1234-abcd-1234567890ab"
-        Disables the server with the specified UUID.
-    .EXAMPLE
-        Get-XoServer | Where-Object { $_.Enabled } | Disable-XoServer
-        Disables all enabled servers.
-    #>
-    [CmdletBinding(SupportsShouldProcess, ConfirmImpact = "Medium")]
-    param(
-        [Parameter(Mandatory, ValueFromPipelineByPropertyName)]
-        [string[]]$ServerUuid
-    )
-    
-    process {
-        foreach ($id in $ServerUuid) {
-            if ($PSCmdlet.ShouldProcess($id, "Disable server")) {
-                try {
-                    Write-Verbose "Disabling server $id"
-                    Invoke-RestMethod -Uri "$script:XoHost/rest/v0/servers/$id/disable" -Method Post @script:XoRestParameters
-                    Write-Verbose "Server disabled successfully"
-                }
-                catch {
-                    Write-Error "Failed to disable server $id. Error: $_"
-                }
-            }
-        }
-    }
-}
```
Why were these cmdlets removed?

iButcat replied:
As explained in your comments in your fork. I added them but since it was an alpha those weren't exposed as cmdlets. They had some unexpected behavior and I needed to reduce the volume of the first version.

dinhngtu commented on src/sr.ps1:
```diff
@@ -29,10 +29,13 @@ function Get-XoSr {
         The ID(s) of the SR(s) to retrieve.
     .PARAMETER Limit
-        Maximum number of results to return when retrieving all SRs.
+        Maximum number of results to return. Default is 25 if not specified.
```
Same: this wants to be presettable per-session

dinhngtu commented on src/server.ps1:
```diff
@@ -111,137 +116,25 @@ function Get-XoServer {
-                            Write-Warning "Failed to process server from URL $serverUrl. Error: $_"
+                            throw "Failed to process server from URL $serverUrl. Error: $_"
                         }
                     }
```
No need for so many try/catch levels, just throw whenever you get an error.

dinhngtu commented on src/vm.ps1:
```diff
@@ -57,74 +68,114 @@ function Get-XoVm {
-            
-            if ($Tag) {
-                $filterParts += "tags:|($($Tag -join ' '))"
+            if ($PowerState) { $filterParts += "power_state:|($($PowerState -join ' '))" }
+            if ($Tag) { $filterParts += "tags:|($($Tag -join ' '))" }
+            if ($Filter) { $filterParts += $Filter }
+            $combinedFilter = if ($filterParts.Count -gt 0) { $filterParts -join " " } else { $null }
+
+            $queryParams = @{
+                fields = $script:XO_VM_FIELDS
             }
```
The filters want processing in `begin{}`.

dinhngtu replied:
never mind, misread the code

iButcat committed:
fix comments, follow same code style, fix limit, id, list id, etc...

dinhngtu commented on src/host.ps1:
```diff
@@ -1,43 +1,97 @@
-            ProductBrand = $InputObject.software_version.product_brand
+    try {
+        $uri = "$script:XoHost/rest/v0/hosts/$HostId"
+        $hostData = Invoke-RestMethod -Uri $uri @script:XoRestParameters -Body ($Params | ConvertTo-Json -Compress) -Method Get
+        
+        $hostData = $hostData | ConvertFrom-Json -AsHashTable
```
`Invoke-RestMethod` already returns a parsed JSON object, why the conversion? Also, I'm pretty sure that's not the right way to pass parameters to the JSON request.

dinhngtu commented on src/host.ps1:
```diff
@@ -1,43 +1,97 @@
         
-        return $resultObject
+        try {
+            $response = Invoke-RestMethod -Uri $hostDetailUri @script:XoRestParameters -Body ($detailParams | ConvertTo-Json -Compress) -Method Get
+            return $response | ConvertFrom-Json -AsHashTable
```
idem.

dinhngtu commented on src/host.ps1:
```diff
@@ -75,88 +132,59 @@ function Get-XoHost {
-                    Write-Error "Failed to retrieve host with ID $id. Error: $_"
-                }
-            }
+    if ($PSCmdlet.ParameterSetName -eq "HostId") {
+        foreach ($id in $HostId) {
+            Get-XoSingleHostById -HostId $id -Params $params
         }
+        return
     }
```
Unless you put this loop in a process block, it will not work with pipelines.

dinhngtu commented on src/host.ps1:
```diff
@@ -75,88 +132,59 @@ function Get-XoHost {
         [Parameter(ParameterSetName = "Filter")]
         [Parameter(ParameterSetName = "All")]
-        [int]$Limit
+        [int]$Limit = $(if ($null -ne $script:XO_DEFAULT_LIMIT) { $script:XO_DEFAULT_LIMIT } else { 25 })
```
Can you set this directly to `$script:XO_DEFAULT_LIMIT`?

dinhngtu commented on src/host.ps1:
```diff
@@ -75,88 +132,59 @@ function Get-XoHost {
+            return
+        }
+
+        $hostsToProcess = $hostsResponse
+        if ($Limit -gt 0 -and $hostsResponse.Count -gt $Limit) {
+            $hostsToProcess = $hostsResponse[0..($Limit-1)]
+        }
+
+        foreach ($hostDetail in $hostsToProcess) {
+            $fullHostDetail = Get-XoHostDetailFromPath -HostDetail $hostDetail
+            if ($fullHostDetail) {
+                $hostObj = ConvertTo-XoHostObject -InputObject $fullHostDetail
+                Write-Output $hostObj
             }
         }
```
@iButcat Again, why query every item in a loop instead of passing the field list and limit 
directly? With this you will generate one JSON request per command + another request for every single object fetched.

dinhngtu commented on src/server.ps1:
```diff
@@ -56,192 +131,62 @@ function Get-XoServer {
+            return
+        }
+        
+        $serversToProcess = $serversResponse
+        if ($Limit -gt 0 -and $serversResponse.Count -gt $Limit) {
+            $serversToProcess = $serversResponse[0..($Limit-1)]
+        }
+        
+        foreach ($serverUrl in $serversToProcess) {
+            $serverDetail = Get-XoServerDetailFromUrl -ServerUrl $serverUrl
+            if ($serverDetail) {
+                $serverObj = ConvertTo-XoServerObject -InputObject $serverDetail
+                Write-Output $serverObj
             }
         }
```
idem.

dinhngtu commented on src/session.ps1:
```diff
@@ -160,3 +161,42 @@ function Disconnect-XoSession {
+        [int]$Limit
+    )
+
+    # Initialize the variable if it doesn't exist
+    if ($null -eq $script:XO_DEFAULT_LIMIT) {
+        $script:XO_DEFAULT_LIMIT = 25
+        Write-Verbose "Default limit was not initialized, setting initial value to 25"
+    }
+
+    # Set the new limit
+    $oldLimit = $script:XO_DEFAULT_LIMIT
+    $script:XO_DEFAULT_LIMIT = $Limit
+    
```
`$script:XO_DEFAULT_LIMIT` should be a constant that's not touched by user commands. The default session limit should be named e.g. `$script:XoLimit` or `$script:XoSessionLimit`. It should also be reset on disconnection.

dinhngtu commented on src/sr.ps1:
```diff
@@ -43,91 +103,64 @@ function Get-XoSr {
-                }
-                else {
-                    Write-Verbose "No SRs found"
+        $srsToProcess = $response
+        if ($Limit -gt 0 -and $response.Count -gt $Limit) {
+            $srsToProcess = $response[0..($Limit-1)]
+        }
+        
+        foreach ($srPath in $srsToProcess) {
+            if ($srPath -is [string]) {
+                $srDetail = Get-XoSrDetailFromPath -SrPath $srPath
+                if ($srDetail) {
+                    ConvertTo-XoSrObject -InputObject $srDetail
```
idem.

dinhngtu commented on src/task.ps1:
```diff
@@ -1,76 +1,258 @@
+        
+        Write-Verbose "Found $($response.Count) tasks"
+        
+        # Apply limit if needed
+        $tasksToProcess = $response
+        if ($Limit -gt 0 -and $response.Count -gt $Limit) {
+            $tasksToProcess = $response[0..($Limit-1)]
+        }
+        
+        foreach ($taskUrl in $tasksToProcess) {
+            if ($taskUrl -is [string] -and $taskUrl -match '/tasks/([0-9a-z]+)') {
+                $id = $matches[1]
+            
+                Get-XoSingleTaskById -TaskId $id -Params $params
+            }
+            else {
+                if ($taskUrl -is [PSObject] -or $taskUrl -is [Hashtable]) {
+                    ConvertTo-XoTaskObject -InputObject $taskUrl
+                }
+                else {
+                    Write-Verbose "Skipping unknown task format: $taskUrl"
+                }
+            }
         }
```
idem.

dinhngtu commented on src/vdi-snapshot.ps1:
```diff
@@ -18,14 +18,53 @@ function ConvertTo-XoVdiSnapshotObject {
+    if ($Item -is [string] -and $Item -match '\/vdi-snapshots\/([^\/]+)(?:$|\?)') {
+        return $matches[1]
+    }
+    elseif ($Item.PSObject.Properties.Name -contains 'id') {
+        return $Item.id
+    }
+    elseif ($Item.PSObject.Properties.Name -contains 'uuid') {
+        return $Item.uuid
+    }
```
Again, is there a difference between `id` and `uuid`? Let's stick with `uuid`.

dinhngtu commented on src/vdi-snapshot.ps1:
```diff
@@ -49,104 +88,68 @@ function Get-XoVdiSnapshot {
-            }
-            catch {
-                Write-Error "Failed to retrieve VDI snapshots. $_"
+        $snapshotsToProcess = $response
+        if ($Limit -gt 0 -and $response.Count -gt $Limit) {
+            $snapshotsToProcess = $response[0..($Limit-1)]
+        }
+
+        foreach ($item in $snapshotsToProcess) {
+            $id = Get-XoVdiSnapshotIdFromItem -Item $item
+            if ($id) {
+                Get-XoSingleVdiSnapshotById -SnapshotId $id -Params $params
```
idem.

dinhngtu commented on src/vdi.ps1:
```diff
@@ -95,74 +130,69 @@ function Get-XoVdi {
+
+        $vdisToProcess = $response
+        if ($Limit -gt 0 -and $response.Count -gt $Limit) {
+            $vdisToProcess = $response[0..($Limit-1)]
+        }
+
+        foreach ($item in $vdisToProcess) {
+            if ($item -is [System.Management.Automation.PSObject] -or $item -is [Hashtable]) {
+                ConvertTo-XoVdiObject $item
             }
-            catch {
-                Write-Error "Failed to retrieve VDIs. Error: $_"
+            else {
+                $id = Get-XoVdiIdFromItem -Item $item
+                if ($id) {
+                    Get-XoSingleVdiById -VdiId $id -Params $params
+                }
```
idem.

dinhngtu commented on src/host.ps1:
```diff
@@ -75,88 +132,59 @@ function Get-XoHost {
-                Write-Error "Failed to list hosts. Error: $_"
+    try {
+        $uri = "$script:XoHost/rest/v0/hosts"
+        $hostsResponse = Invoke-RestMethod -Uri $uri @script:XoRestParameters -Body ($params | ConvertTo-Json -Compress) -Method Get
+        
+        if (!$hostsResponse -or $hostsResponse.Count -eq 0) {
+            Write-Verbose "No hosts found"
+            return
+        }
```
idem for request body. Also, why the check and return here?

dinhngtu commented on src/vdi-snapshot.ps1:
```diff
@@ -49,104 +88,68 @@ function Get-XoVdiSnapshot {
-                            $id = $null
+    try {
+        Write-Verbose "Getting all VDI snapshots with parameters: $($params | ConvertTo-Json -Compress)"
+        $response = Invoke-RestMethod -Uri "$script:XoHost/rest/v0/vdi-snapshots" @script:XoRestParameters -Body ($params | ConvertTo-Json -Compress) -Method Get
```
idem: body params

dinhngtu commented on src/vdi.ps1:
```diff
@@ -95,74 +130,69 @@ function Get-XoVdi {
+        }
+    }
+
+    if ($PSCmdlet.ParameterSetName -eq "VdiId") {
```
Why do some object types use XxxId for UUID parameters and some other use XxxUuid?

dinhngtu commented:
Overall comments:

- You must make sure that you're passing parameters to your REST requests correctly (using a hashtable in `-Body`). Similarly, `Invoke-RestMethod` already returns a parsed object, don't parse it further with `ConvertFrom-Json`.
- Object ID field names should be consistent: `HostUuid`, `SrUuid`, etc. (`TaskId` is like that because task IDs are not UUIDs.) Alternatively, you can convert everything to `HostId`/etc.; but let's be consistent with the naming. However, don't accept `id` or `uuid` as parameter names or parameter aliases. Any object you pipe in that contains these fields will be used for parameter name matching, which will cause unexpected behavior.
- The functions for querying objects from URLs are all over the place, `Get-XoHostDetailFromPath`, `Get-XoServerDetailFromUrl`, `Get-XoVdiSnapshotIdFromItem`, `Get-XoVdiIdFromItem`, etc. Each function has a different structure.
- In your main get cmdlets (e.g. `Get-XoSr`), you must use `begin/process/end{}` to separate the per-ID and filter-based queries. Otherwise, pipelines won't work correctly (think piping a list of objects with SrId to Get-XoSr)

dinhngtu commented:
You should break your changes into separate commits for each functionality change. Also, please sign off all of your commits.

iButcat commented:
Thanks for the review, let me fix that directly.

iButcat committed:
Fix Get-XoVm and Get-XoVmSnapshot to avoid individual API calls

- Removed unnecessary API calls in the foreach loop when processing responses
- Added proper limit handling for returned data
- Simplified snapshot retrieval process to use one API call with proper fields
- Fixed parameter aliases for consistent naming convention

iButcat committed:
Fix Get-XoTask parameter defaults and response handling

- Removed the default status filter to allow fetching all tasks
- Added proper conditional handling for status filtering
- Fixed verbose output messages to be clearer about filtering
- Updated response processing to handle different API response formats

iButcat committed:
Fix response processing and parameter naming consistency

- Ensured all response handlers convert objects properly
- Fixed parameter naming to use *Uuid consistently for UUID parameters
- Added proper aliases for backward compatibility
- Removed redundant string type checks

dinhngtu commented on src/host.ps1:
```diff
@@ -1,43 +1,100 @@
+        }
     }
+    return $HostPath
 }
```
This function is never used. There are several other similar functions that are also unused.

dinhngtu commented on src/host.ps1:
```diff
@@ -47,116 +104,89 @@ function Get-XoHost {
+                $hostsToProcess = $hostsResponse
+                if ($Limit -gt 0 -and $hostsResponse.Count -gt $Limit) {
+                    $hostsToProcess = $hostsResponse[0..($Limit-1)]
                 }
```
Why filter the results array here if `params` already contains a limit factor?

iButcat replied:
Indeed. I removed it

dinhngtu commented on src/host.ps1:
```diff
@@ -47,116 +104,89 @@ function Get-XoHost {
         [Parameter(ParameterSetName = "Filter")]
         [Parameter(ParameterSetName = "All")]
-        [int]$Limit
+        [int]$Limit = $script:XoSessionLimit
```
@iButcat What's the difference between the "Filter" and "All" parameter sets?

dinhngtu commented on src/server.ps1:
```diff
@@ -1,247 +1,184 @@
 
     process {
         $props = @{
             ServerUuid = $InputObject.id
```
Doesn't this want to be `uuid` and not `id`?

iButcat replied:
Not for server, only an id is returned from the api response. See screenshots attached. 

For example host indeed contains both uuid & id. 

Host example 
<img width="384" alt="Screenshot 2025-04-11 at 16 19 27" src="https://github.com/user-attachments/assets/444ea513-5e69-4d81-8556-cfa1f971cc40" />

Server example:
<img width="441" alt="Screenshot 2025-04-11 at 16 20 35" src="https://github.com/user-attachments/assets/8ae16ac1-2e75-4190-ac99-4f97d3c3558f" />

dinhngtu commented on src/vdi.ps1:
Why the path resolution? All cmdlets work just fine with PS paths.

dinhngtu reviewed (approved):
Please do the following:

- Fix all merge conflicts (these are just the license identifiers, you can just prepend `# SPDX-License-Identifier: Apache-2.0` at the top of your ps1 files);
- Sign off **all** of your commits in the PR.
- Address a few pending comments.

You can go ahead and merge once this is done.

iButcat committed:
refactor: remove unused DetailFromPath/URL functions and standardize on Filter parameter set

- Removed unused Get-Xo*DetailFromPath/URL functions
- Removed Get-XoVdiIdFromItem and Get-XoVdiSnapshotIdFromItem
- Changed default parameter set from All to Filter
- Added comments explaining parameter sets

Signed-off-by: Butcat <butcat@protonmail.com>

iButcat committed:
fix: remove redundant client-side filtering

- Removed unnecessary client-side filtering in Get-XoHost, Get-XoServer, Get-XoSr, Get-XoVm, and Get-XoTask
- Rely on API's limit parameter for server-side filtering

Signed-off-by: Butcat <butcat@protonmail.com>

iButcat committed:
Add session limit management

- Add  variable
- Reset limit on connect/disconnect
- Update Set-XoDefaultLimit to use new variable

Signed-off-by: Butcat <butcat@protonmail.com>

iButcat commented:
Fix issue #1
